### PR TITLE
Add PostCSS plugin to fix relative `@content` and `@plugin` paths in `@import`ed files

### PR DESCRIPTION
**Pull request by philipp-spiess**
We noticed an issue that happened when handling relative file imports in the `@plugin` and the upcoming `@content` APIs. The problem arises from relative files that are inside `@import`ed stylesheets. Take, for example, the following folder structure:

```css
/* src/index.css */
@import "./dir/index.css";
```
```css
/* src/dir/index.css */
@plugin "../../plugin.ts";
```

It's expected that the path is relative to the CSS file that defined it. However, right now, we use [`postcss-import`](https://github.com/postcss/postcss-import) to flatten the CSS file before running the tailwind build step. This causes these custom-properties to be inlined in a flat file which removes the information of which file is being referred:

```css
/* src/flat.css */
@plugin "../../plugin.ts"; /* <- This is now pointing to the wrong file */
```

There are generally two approaches that we can do to solve this:

1. **Handle `@import` flattening inside tailwindcss:** While generally this would give us more freedom and less dependencies, this would require some work to get all edge cases right. We need to support layers/conditional imports and also handle all relative urls for properties like `background-image`. 
2. **Rewrite relative paths as a separate postcss visitor:** The approach this PR takes is instead to implement a custom postcss plugin that uses the AST to rewrite relative references inside `@plugin` and `@content`. This has the benefit of requiring little changes to our existing APIs. The rule is only enabled for relative references inside `@plugin` and `@content`, so the surface of this rule is very small.

We can use this plugin inside all three current clients: 

- `@tailwindcss/postcss` obviously already uses postcss
- `@tailwindcss/cli` also uses postcss to handle `@import` flattening
- `@tailwindcss/vite` allows us to add custom postcss rules via the CSS pipeline. There are a few cases that we handle with care (e.g. in vite you can pass a string to the postcss config which is supposed to load the config from a file).

## Test plan

To validate the changes, we have added both a list of unit test cases to the plugin itself as well as verified that all three clients are working as expected:

- `@tailwindcss/postcss` now has an explicit test for this behavior
- `@tailwindcss/cli` and `@tailwindcss/vite` were manually tested by updating the vite playground. The CLI was run with `--cwd playgrounds/vite/ -i ./src/app.css -o foo.css`: 
    Screenshot 2024-07-29 at 11 35 59

### CODE REVIEW

File: packages/postcss-fix-relative-paths-plugin/src/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import fs from 'node:fs'
+import postcss from 'postcss'
+import atImport from 'postcss-import'
+import { describe, expect, test } from 'vitest'
+import fixRelativePathsPlugin from '.'
+
+describe('fixRelativePathsPlugin', () => {
+  test('rewrites @content and @plugin to be relative to the initial css file', async () => {
+    const CSS_PATH = `${__dirname}/fixtures/external-import/src/index.css`
```
Comment (philipp-spiess) [inline packages/postcss-fix-relative-paths-plugin/src/index.test.ts:9]: 🫠

File: packages/@tailwindcss-postcss/src/index.ts [hunk 2]
```diff
@@ -48,114 +49,119 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
     }
   })
 
+  let hasApply = false
+  let hasTailwind = false
```
Comment (adamwathan) [inline packages/@tailwindcss-postcss/src/index.ts:53]: Would it be clearer if we initialized these without a value so it doesn't seem like these `false` values are actually important to how the code works?

`let hasApply, hasTailwind;`

File: packages/internal-postcss-fix-relative-paths/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import type { AtRule, Container, Plugin } from 'postcss'
+
+const SINGLE_QUOTE = "'"
+const DOUBLE_QUOTE = '"'
+
+export default function fixRelativePathsPlugin(): Plugin {
+  // Retain a list of touched at-rules to avoid infinite loops
+  let touched: WeakSet<AtRule> = new WeakSet()
+
+  function fixRelativePath(atRule: AtRule, helpers: any) {
```
Comment (adamwathan) [inline packages/internal-postcss-fix-relative-paths/src/index.ts:11]: Looks like `helpers` isn't used anymore 👍

File: packages/internal-postcss-fix-relative-paths/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+[Discuss Tailwind CSS on GitHub](https://github.com/tailwindcss/tailwindcss/discussions)
+
+For chatting with others using the framework:
+
+[Join the Tailwind CSS Discord Server](https://discord.gg/7NF8GNe)
+
+## Contributing
+
+If you're interested in contributing to Tailwind CSS, please read our [contributing docs](https://github.com/tailwindcss/tailwindcss/blob/next/.github/CONTRIBUTING.md) **before submitting a pull request**.
```
Comment (adamwathan) [inline packages/internal-postcss-fix-relative-paths/README.md:40]: Don't think we need a README for internal packages like this one 👍

**Review by adamwathan — APPROVED** [review]

**Pull request merged**